resource-managers/yarn/src/main/scala/org/apache/spark/deploy/yarn/ApplicationMaster.scala

```diff
@@ -863,10 +864,20 @@ object ApplicationMaster extends Logging {
     val ugi = sparkConf.get(PRINCIPAL) match {
       // We only need to log in with the keytab in cluster mode. In client mode, the driver
       // handles the user keytab.
-      case Some(principal) if amArgs.userClass != null =>
+      case Some(principal) if master.isClusterMode =>
         val originalCreds = UserGroupInformation.getCurrentUser().getCredentials()
         SparkHadoopUtil.get.loginUserFromKeytab(principal, sparkConf.get(KEYTAB).orNull)
         val newUGI = UserGroupInformation.getCurrentUser()
+
+        // Set the context class loader so that the token manager has access to jars
+        // distributed by the user.
+        Utils.withContextClassLoader(master.userClassLoader) {
+          // Re-obtain delegation tokens, as they might be outdated as of now. Add the fresh
+          // tokens on top of the original user's credentials (overwrite).
+          val credentialManager = new HadoopDelegationTokenManager(sparkConf, yarnConf, null)
+          credentialManager.obtainDelegationTokens(originalCreds)
```

Comment:
This is unnecessary for the first attempt, since you should have fresh tokens from spark-submit. But not a big deal either.

Reply:
Technically yes, but I didn't because I have no idea how to manually check whether the tokens are expired or not. I think that's not exposed in common interface for token.

Reply:
Ah you meant "attempt". Thanks! Will address.